### PR TITLE
Adds a tooltip to the transaction tables to show the imported payee

### LINKED ISSUE #782: [Feature] Allow viewing the imported payee
### Verified feature request does not already exist?

- [X] I have searched and found no existing issue

### 💻

- [ ] Would you like to implement this feature?

### Pitch: what problem are you trying to solve?

We store the imported payee for transactions but it is presently not visible anywhere on the transaction (outside of the rules modal).

### Describe your ideal solution to this problem

Add some sort of interaction to show the imported payee — maybe a tooltip? Or a button somewhere to toggle the payee column to “imported payee?”

### Teaching and learning

_No response_

### PR DESCRIPTION
**Pull request by matt-fidd**
Putting this up for PR to get some opinions. (Fixes #782)

I think a way to view the imported payee (other than creating a rule from a transaction) would be hugely beneficial and, to me, this is cleaner than another column in the table.

I'm not sure if it would be obvious to people what it was showing so maybe it needs an `Imported Payee: ${payeeName}` in the tooltip instead.

Anyway, let me have your thoughts please! 😄 

To test, import a budget file with some imported transactions, then hover over the payee field in the table.

Based on #2828 so a lot of the diff is unrelated.

### CODE REVIEW

**Comment by youngcw** [comment]
How does one view this tooltip?

**Comment by matt-fidd** [comment]
> How does one view this tooltip?

Sorry, just updated the PR details. 
To test, import a budget file with some imported transactions, then hover over the payee field in the table.

**Comment by youngcw** [comment]
I would say that the tooltip should be bigger, and it probably would help to add in "imported payee" to it, like you were saying.  The delay to show could maybe be a touch longer too, but thats not a big deal right now

**Comment by matt-fidd** [comment]
Upped the delay to 750ms which feels a bit better, and added a bit of padding.

I've got two options for the layout:
1:
image

2:
image

There's also the option for tooltip placement, I've changed it to pop out from the right instead of the top but I'm not sure if I prefer it.

**Comment by youngcw** [comment]
I prefer the second option.

**Comment by youngcw** [comment]
Would it be worth adding in some sort of indicator that there is an imported_payee that can be shown?

**Comment by matt-fidd** [comment]
It could be useful, but I don't want to clutter the table. Do you have any suggestions as to how it might look?

**Comment by youngcw** [comment]
I agree that if there is anything it should be really simple and not intrusive.  Maybe like a subdued icon of some kind.  But that also could be for a later PR.

**Comment by psybers** [comment]
What about a dotted underline?  Usually dotted underlines mean there is more info/hovers available. Solid underlines mean link.

Though I think the dotted underlines usually require you to still click them to get the hover and it is not automatic.  Still might work in this case?

**Comment by psybers** [comment]
Can we avoid showing the imported payee if it is identical to the current payee?

**Comment by psybers** [comment]
I also wonder if the tooltip placement would not be better under the payee name, instead of to the right of it.  That would allow for more easily comparing the two.

**Comment by psybers** [comment]
https://ux.stackexchange.com/questions/133212/standard-style-for-text-that-opens-hover-tool-tip

Something like this? https://developer.mozilla.org/en-US/docs/Web/HTML/Element/abbr#providing_an_expansion

File: upcoming-release-notes/2828.md
Comment (psybers) [inline upcoming-release-notes/2828.md:1]: Not sure what happened here, but you seem to have two PRs in here? There are unrelated changes from #2828 in here.

**Review by psybers — CHANGES_REQUESTED** [review]

File: upcoming-release-notes/2828.md
Comment (matt-fidd) [reply upcoming-release-notes/2828.md:1]: I mentioned this in my comments, #2828 migrated this file to the new Tooltip/Popover react-aria so it's based on top of that branch. Once it's merged in I'll rebase to master

**Comment by matt-fidd** [comment]
> I also wonder if the tooltip placement would not be better under the payee name, instead of to the right of it. That would allow for more easily comparing the two.

You're right, done

**Comment by matt-fidd** [comment]
> What about a dotted underline? Usually dotted underlines mean there is more info/hovers available. Solid underlines mean link.
> 
> Though I think the dotted underlines usually require you to still click them to get the hover and it is not automatic. Still might work in this case?

Good idea! That's subtle enough and doesn't make it look too busy. Try with your own file if you can to get a better idea

image

**Comment by matt-fidd** [comment]
> Can we avoid showing the imported payee if it is identical to the current payee?

I'm not sure on this, I personally prefer it always-on.

**Comment by psybers** [comment]
It just seems odd to see things like this:

image

You could easily do this by changing the `importedPayee ?` guard to be `importedPayee && payee != importedPayee ?`.

**Comment by youngcw** [comment]
I feel like seeing that they are the same could be useful.  Like "these are the same.  Maybe I should make a real payee for this"

**Comment by MatissJanis** [comment]
(updated the base branch so it's easier to see the real diff)

**Comment by matt-fidd** [comment]
Yes, you're probably both right. I think it made less sense to me because I have very few importedPayees the same as payee.

Updated, let me know what you think

**Comment by matt-fidd** [comment]
The netlify build appears to be struggling

**Comment by youngcw** [comment]
If the original problem this PR was trying to solve was to make the imported payee visible in the gui, doesn't selectively not showing it kinda defeat the whole point?  Like there is no way to know if a transaction actually has an imported payee since there could be one, but its just not showing.  There is no way to know without resorting to the original workarounds we were trying to not need.

**Comment by psybers** [comment]
@youngcw What problem(s) was being solved with this PR? The original issue was to be able to view the imported payee.  Even with selectively applying the tooltips, you are viewing the imported payee (because it is showing in the payee column).  If you then change the payee to something else, it would show the imported payee.

What use case(s) are enabled by not selectively filtering?

**Comment by youngcw** [comment]
Im thinking about the use case of trying to make some rules based on imported payee, but you want to see what that imported payee is first on a few transactions (like similar but slightly different imported payee).  

It looks like, as of now, that if the imported payee and the payee name match, then there is no indication that an imported payee exists at all.  You would have to know before hand that the imported payee and the payee happen to be matching. (Often its obvious because of weird formatting, but thats not always the case)  So if you didn't know beforehand that they were the same, and see that the payee doesn't have the option of a tooltip, the expected meaning of that would be that there is no imported payee.

Another side use case would be if you wanted to know if a transaction came from import or manual addition, the indicator of imported payee would be a quick visual of that.

**Comment by psybers** [comment]
> > The original issue was to be able to view the imported payee
> 
> I agree that this is the goal: to be able to view the imported payee. And wouldn't hiding the fact that there is one defeat this goal?

I guess that is all in how you interpret the semantics of the request.  I interpret it as being able to view the imported payee.  Which for these cases, is already being displayed in the payee column. Hence, it is already fulfilled for those cases and we should focus on the others.

> Another side use case would be if you wanted to know if a transaction came from import or manual addition, the indicator of imported payee would be a quick visual of that.

Glad you mentioned this second case!  I think we need indicators that show if it was manually entered, imported via bank sync, or added via API.

**Comment by youngcw** [comment]
@matt-fidd I don't know why this specific PR is having issues with netlify.  The others seem fine.

**Comment by matt-fidd** [comment]
I wonder if the netlify issue is because the workflow isn't being run because of the different base branch. I can see that it might just target merges into `master`.

**Comment by youngcw** [comment]
> I wonder if the netlify issue is because the workflow isn't being run because of the different base branch. I can see that it might just target merges into `master`.

Lets find out.  Can you try pushing some dummy commit?

**Comment by matt-fidd** [comment]
Yes, that's sorted it.

**Comment by matt-fidd** [comment]
With #2828 merged, this is back open for a proper review.

File: packages/desktop-client/src/components/transactions/TransactionsTable.jsx [hunk 1]
```diff
@@ -562,17 +564,46 @@ function PayeeCell({
...
+                transaction={transaction}
+                transferAccount={transferAcct}
+                onNavigateToTransferAccount={onNavigateToTransferAccount}
+                onNavigateToSchedule={onNavigateToSchedule}
+              />
+              <UnexposedCellContent
+                {...props}
+                style={{ borderBottom: `1px dashed ${theme.pageTextSubdued}` }}
+              />
+            </div>
```
Comment (psybers) [inline packages/desktop-client/src/components/transactions/TransactionsTable.jsx:593]: Is it possible to pull this up into it's own variable, to avoid repeating it below?  The only difference is the `style` on the `UnexposedCellContent`, which you could easily set with a ternary.

Then the code would look more like `importedPayee ? <tooltip>$var</tooltip> : $var`.

**Review by psybers — CHANGES_REQUESTED** [review]

File: packages/desktop-client/src/components/transactions/TransactionsTable.jsx (hunk 1 above)
Comment (matt-fidd) [reply packages/desktop-client/src/components/transactions/TransactionsTable.jsx:593]: I'm not ignoring this, it _can't_ be hard but I appear to be fighting with React. Will give it another go with a fresh head tomorrow. I think it's something props related.

Comment (matt-fidd) [reply packages/desktop-client/src/components/transactions/TransactionsTable.jsx]: Ha! Got it! Updated now.

**Review by psybers — APPROVED** [review]

**Comment by matt-fidd** [comment]
Seem to be having some trouble with split transactions and upcoming schedules, the flex grow
image

is causing the border to extend
image

**Comment by matt-fidd** [comment]
It looks like the transactions that are behaving have an extra div from somewhere. (.css-83enl9 in this case)

image

Does anyone know how I can track this down? I can't see any obvious difference between a child and a normal transaction. Also happening with scheduled transactions that have not been posted.

**Comment by matt-fidd** [comment]
In VRT, they're all coming up full width.

image

**Comment by youngcw** [comment]
I wonder if the vrt issues is because of a smaller screen size?

File: packages/desktop-client/src/components/transactions/TransactionsTable.jsx [hunk 2]
```diff
@@ -562,17 +564,40 @@ function PayeeCell({
...
+          <div style={{ display: 'flex', alignItems: 'center', width: '100%' }}>
+            <PayeeIcons
+              transaction={transaction}
+              transferAccount={transferAcct}
+              onNavigateToTransferAccount={onNavigateToTransferAccount}
+              onNavigateToSchedule={onNavigateToSchedule}
+            />
+            <UnexposedCellContent
+              {...props}
+              style={{ borderBottom: `1px dashed ${theme.pageTextSubdued}` }}
```
Comment (psybers) [inline packages/desktop-client/src/components/transactions/TransactionsTable.jsx:578]: Shouldn't this only be underlined if there is an importedPayee?

**Review by psybers — CHANGES_REQUESTED** [review]

File: packages/desktop-client/src/components/transactions/TransactionsTable.jsx [hunk 3]
```diff
@@ -562,17 +564,40 @@ function PayeeCell({
...
+            />
+          </div>
+        );
+
+        return importedPayee ? (
+          <Tooltip
+            content={
+              <View style={{ padding: 10 }}>
+                <Text style={{ fontWeight: 'bold' }}>Imported Payee</Text>
+                <Text style={{ fontWeight: 'normal' }}>{importedPayee}</Text>
```
Comment (psybers) [inline packages/desktop-client/src/components/transactions/TransactionsTable.jsx:588]: Is the normal needed here, shouldn't that already be the case?
```suggestion
                <Text>{importedPayee}</Text>
```

File: packages/desktop-client/src/components/transactions/TransactionsTable.jsx [hunk 4]
```diff
@@ -562,17 +564,40 @@ function PayeeCell({
...
-            transferAccount={transferAcct}
-            onNavigateToTransferAccount={onNavigateToTransferAccount}
-            onNavigateToSchedule={onNavigateToSchedule}
-          />
-          <UnexposedCellContent {...props} />
-        </>
-      )}
+      unexposedContent={props => {
+        const payeeInfo = (
+          <div style={{ display: 'flex', alignItems: 'center', width: '100%' }}>
```
Comment (psybers) [inline packages/desktop-client/src/components/transactions/TransactionsTable.jsx:569]: Is this the VRT issue?  You set the width to 100% here so the children will fill it?

**Review by psybers — CHANGES_REQUESTED** [review]

File: packages/desktop-client/src/components/transactions/TransactionsTable.jsx (hunk 4 above)
Comment (matt-fidd) [reply packages/desktop-client/src/components/transactions/TransactionsTable.jsx:569]: Most likely, I have no idea why I did that... must have been a reason. Will check when I can get back to a dev box.

File: packages/desktop-client/src/components/transactions/TransactionsTable.jsx (hunk 3 above)
Comment (matt-fidd) [reply packages/desktop-client/src/components/transactions/TransactionsTable.jsx:588]: Yes, think this left over from a refactor where I had them all in the same line and needed to reset.

File: packages/desktop-client/src/components/transactions/TransactionsTable.jsx (hunk 2 above)
Comment (matt-fidd) [reply packages/desktop-client/src/components/transactions/TransactionsTable.jsx:578]: https://github.com/actualbudget/actual/pull/3018#issuecomment-2222878509

**Comment by matt-fidd** [comment]
I'll update VRT when I get home after work.

**Comment by youngcw** [comment]
Im seeing inconsistent lengths of the underline.  Some full box some right length.  Im also seeing that many of the schedule icons are getting misaligned and appear above the payee, but not all of them weirdly.

**Comment by matt-fidd** [comment]
Netlify mustn't have updated for me properly, it looked fine a few minutes ago but I can see the same issues now.
That must have been the reason for the full width flex.

File: packages/desktop-client/src/components/transactions/TransactionsTable.jsx (hunk 2 above)
Comment (matt-fidd) [reply packages/desktop-client/src/components/transactions/TransactionsTable.jsx:578]: Sorry, as I was eating I only just realised what you meant. Will fix.

**Comment by youngcw** [comment]
Looking good!  I noticed that hovering the transfer or schedule icons will show the tooltip too.  Thats not a big problem but it does seem a little weird.

**Comment by psybers** [comment]
> Looking good! I noticed that hovering the transfer or schedule icons will show the tooltip too. Thats not a big problem but it does seem a little weird.

Probably my fault for suggesting lifting the common code into a single place. Now the tooltip wraps everything...

**Comment by matt-fidd** [comment]
Fixed now, I've moved the icons out of the tooltip

**Comment by matt-fidd** [comment]
Updated to fix conflicts from recent parent payee display change.

**Comment by youngcw** [comment]
If the transaction is a split then there doesn't seem to be a way to view the imported payee there.  Am I missing something?

**Comment by matt-fidd** [comment]
It looks like the imported payee information isn't present in a child transaction.

image

I'll have to rewrite it to pull in the info from the parent.

**Comment by matt-fidd** [comment]
Alternatively, should the imported payee be pulled through to the child on creation as it is with payee? I imagine there's a migration that could be written to pull the info through historically if we go in that direction.

**Comment by youngcw** [comment]
Could we still have the option to hover and view it on the parent?  The child payees might be different, plus the child payees weren't the ones that got imported, most of the time.

**Comment by matt-fidd** [comment]
> Could we still have the option to hover and view it on the parent? The child payees might be different, plus the child lines weren't the ones that got imported, most of the time.

Done, it looks a little different because of the split button it's inside but works the same.

**Comment by matt-fidd** [comment]
Do you think the delay is a bit long or about right?

**Comment by youngcw** [comment]
The delay time feels good to me

**Review by youngcw — APPROVED** [review]
